File: flowers_back/app/schemas/product.py
from pydantic import BaseModel, Field
from typing import Optional, List

from app.models.product import ProductAvailabilityVariants
from pydantic import BaseModel
from typing import List, Optional


class ProductAttributeValueResponse(BaseModel):
    attribute_name: str
    value: str


class ProductWithAttributesResponse(BaseModel):
    id: int
    name: str
    description: Optional[str]
    price: float
    ingredients: Optional[str]
    photo_url: Optional[str]
    attributes: List[ProductAttributeValueResponse]


class ProductResponse(BaseModel):
    id: int
    name: str
    description: Optional[str]
    price: float
    ingredients: Optional[str]
    photo_url: Optional[str] = ""
    categoryId: Optional[int] = None,
    availability: ProductAvailabilityVariants = ProductAvailabilityVariants.AVAILABLE

    class Config:
        from_attributes = True


class ProductImagesResponse(BaseModel):
    id: int
    name: str
    description: Optional[str]
    price: float
    ingredients: Optional[str]
    images: List[str] = []
    categoryId: Optional[int] = None,
    availability: ProductAvailabilityVariants = ProductAvailabilityVariants.AVAILABLE

    class Config:
        from_attributes = True


class ProductPageResponse(BaseModel):
    total: int
    page: int
    per_page: int
    products: List[ProductImagesResponse]

File: flowers_front/src/api/product.ts
import axios from "./core";

export interface IProduct {
    id: number;
    name: string;
    description: string | null;
    price: number;
    ingredients: string;
    images: string[];
    availability: string;
    categoryId: number;
}

export interface IProductBack {
    id: number;
    name: string;
    description: string | null;
    price: number;
    ingredients: string;
    images: string[];
    availability: string;
    categoryId: number;
}


export interface ProductPageResponse {
    total: number;
    page: number;
    perPage: number;
    products: IProduct[];
}

export interface IProductPageResponse {
    total: number;
    page: number;
    perPage: number;
    products: IProductBack[];
}

export const fetchProducts = async (
    page: number = 1,
    perPage: number = 30,
    categoryId?: number | null
): Promise<ProductPageResponse> => {
    try {
        const params: Record<string, any> = { page, per_page: perPage };
        if (categoryId) {
            params.category_id = categoryId;
        }

        const response = await axios.get<IProductPageResponse>("/products/", {
            params,
        });
        return {
            ...response.data,
            products: response.data.products.map((el) => ({
                ...el
            })),
        };
    } catch (error) {
        console.error("Ошибка при загрузке продуктов:", error);
        throw error;
    }
};

export const fetchProductById = async (
    productId: number
): Promise<IProduct> => {
    try {
        const response = await axios.get<IProductBack>(`/products/${productId}`);
        return { ...response.data};
    } catch (error) {
        console.error("Ошибка при загрузке продукта:", error);
        throw error;
    }
};
